File: src/x/crazyjvm/lazyx/s/Lazy.scala
package x.crazyjvm.lazyx.s


/**
 * Created by chenchao on 15/10/12.
 */
class Lazy {

}

object Lazy extends App{
  lazy val x = {
    print("(computing x) ")
    42
  }

  print("x = ")
  println(x)
}

File: src/x/crazyjvm/lazyx/j/Lazy.java
package x.crazyjvm.lazyx.j;

/**
 * Created by chenchao on 15/10/6.
 */

class Component {

}

public class Lazy {
    private volatile Component component;

    public Component getComponent() {
        Component result = component;
        if (result == null) {
            synchronized (this) {
                result = component;
                if (result == null) {
                    component = result = new Component();
                }
            }
        }
        return result;
    }

    public static void main(String[] args) {
    }
}


